File: src/Engine/Engine.cpp
#include "Engine.h"

#include "HumanPlayer.h"
#include "DumbAIPlayer.h"

namespace BeitaGo {
	Engine::Engine() {
		_board = nullptr;
		_player1 = nullptr;
		_player2 = nullptr;
	}

	Engine::~Engine() {
		//TODO: If C++ is nice, does delete just give a nullptr?
		delete _board;
		_board = nullptr;
		delete _player1;
		_player1 = nullptr;
		delete _player2;
	}

	void Engine::NewGame() {
		//TODO: Make these customisable.
		_board = new Board(Grid2(19, 19));
		_player1 = new HumanPlayer(*this, Color::Black);
		_player2 = new DumbAIPlayer(*this, Color::White);
	}

	Board& Engine::GetBoard() {
		return *_board;
	}

	const Board& Engine::GetBoard() const {
		return *_board;
	}

	Player& Engine::GetCurrentPlayer() {
		if (GetPlayer1().GetColor() == GetBoard().GetWhoseTurn()) {
			return GetPlayer1();
		} else {
			return GetPlayer2();
		}
	}

	const Player& Engine::GetCurrentPlayer() const {
		if (GetPlayer1().GetColor() == GetBoard().GetWhoseTurn()) {
			return GetPlayer1();
		} else {
			return GetPlayer2();
		}
	}

	Player& Engine::GetPlayer1() {
		return *_player1;
	}

	const Player& Engine::GetPlayer1() const {
		return *_player1;
	}

	Player& Engine::GetPlayer2() {
		return *_player2;
	}

	const Player& Engine::GetPlayer2() const {
		return *_player2;
	}
}

File: srcpy/engine/engine.py
from .board import Board
from .grid2 import Grid2
# from .player import Player

class Engine:
	def __init__(self):
		self._board = None
		self._player1 = None
		self._player2 = None
	
	def new_game(self, board_dimensions: Grid2, player1: "Player", player2: "Player"):
		self._board = Board(board_dimensions)
		self._player1 = player1
		self._player2 = player2

	def get_board(self) -> Board:
		return self._board

	def get_current_player(self) -> "Player":
		return self._player1 if self.get_board().get_whose_turn() == self._player1.get_color() else self._player2

	def get_player1(self) -> "Player":
		return self._player1

	def get_player2(self) -> "Player":
		return self._player2
